src/channels.py

```diff
@@ -0,0 +1,223 @@
+import time
+
+from src.context import IRCContext, Features
+from src.logger import debuglog
+from src import users
+
+Main = None # main channel
+
+all_channels = {}
+
+_states = ("not yet joined", "pending join", "joined", "pending leave", "left channel", "", "deleted", "cleared")
+
+def _strip(name):
+    return name.lstrip("".join(Features["STATUSMSG"]))
+
+def predicate(name):
+    return not name.startswith(tuple(Features["CHANTYPES"]))
+
+def get(name):
+    """Return an existing channel, or raise a KeyError if it doesn't exist."""
+
+    return all_channels[_strip(name)]
+
+def add(name, cli):
+    """Add and return a new channel, or an existing one if it exists."""
+
+    name = _strip(name)
+
+    if name in all_channels:
+        if cli is not all_channels[name].client:
+            raise RuntimeError("different IRC client for channel {0}".format(name))
+        return all_channels[name]
+
+    cls = Channel
+    if predicate(name):
+        cls = FakeChannel
+
+    chan = all_channels[name] = cls(name, cli)
+    chan.join()
+    return chan
+
+def exists(name):
+    """Return True if a channel with the name exists, False otherwise."""
+
+    return _strip(name) in all_channels
+
+class Channel(IRCContext):
+
+    is_channel = True
+
+    def __init__(self, name, client, **kwargs):
+        super().__init__(name, client, **kwargs)
+        self.users = set()
+        self.modes = {}
+        self.timestamp = None
+        self.state = 0
+
+    def __del__(self):
+        self.users.clear()
+        self.modes.clear()
+        self.state = -2
+        self.client = None
+        self.timestamp = None
+
+    def __str__(self):
+        return "{self.__class__.__name__}: {self.name} ({0})".format(_states[self.state], self=self)
+
+    def __repr__(self):
+        return "{self.__class__.__name__}({self.name!r})".format(self=self)
+
+    def join(self, key=""):
+        if self.state in (0, 4):
+            self.state = 1
+            self.client.send("JOIN {0} :{1}".format(self.name, key))
```

Comment:
Incorrect. Only commands where the argument may contain spaces takes a colon. JOIN is not one of those cases, in fact it's the opposite. JOIN #channel1,#channel2 key1 key2 is the proper syntax. Re-read RFC 1459 if you don't believe me, but JOIN does **not** have a colon in it.